### ISSUE TITLE
Improve accessibility of GUI

### ISSUE THREAD

**Issue by bertfrees** [daisy/pipeline-gui#72]
Reported by Avneesh (tested on Windows):

> I have tried it myself today so I can confirm that it is not such a huge failure as reported by Prashant. But there are some issues that I observed immediately.
> 
> - File menu is not activated by alt+F. Alt key activates the menu but it fails sometimes, especially when the keyboard focus is in a some listbox etc.
> - The browse input text box does not have accessibility label. JAWS just reports textbox and browse button, without saying that it is for selecting input file.
> - Similarly JAWS reports textbox and browse button for output directory without saying that it is for output directory.
> - When I press browse button for output directory, the select directory dialog opens, and while tabbing in this select directory dialog something makes the GUI crash on my machine.
>  
> For verifying accessibility labels, you can install NVDA on your machines and verify yourselves. With this you can ensure basic accessibility while doing coding.

**Comment by bertfrees** [comment]
To do:
- [ ] test the GUI on Windows + NVDA
- [ ] test the GUI on OS X + VoiceOver
- [ ] get in touch with Prashant to figure out the details, and work with him on improving any issue

**Comment by bertfrees** [comment]
OK a quick test with VoiceOver showed that none of the text boxes, dropdown lists, checkboxes, etc. seem to have labels attached.

I'm looking at the GUI code now and from what I can tell there is nothing that explicitly links for example the "Choose a script" title with the scripts dropdown list. I wonder how is this supposed to work? Does JavaFX figure this out itself?

**Comment by bertfrees** [comment]
OK after I saw [this](https://youtu.be/iUPPEkD9H1I?t=20m53s) I tried replacing:
```java
Text title = new Text("Choose a script:");
```

with this:
```java
Label title = new Label("Choose a script:");
title.setLabelFor(scriptsCombo);
```

and that seemed to have made a difference. Now VoiceOver says "Choose a script: pop up button" instead of just "pop up button".

**Comment by bertfrees** [comment]
@rdeltour

**Comment by bertfrees** [comment]
@marisademeglio

**Comment by mccallum-sgd** [comment]
@bertfrees Alright, so I did some testing on my desktop (Windows 10).

**Comment by bertfrees** [comment]
Apart from adding labels to nodes, it is also important to check the following:
- [ ] Some nodes are updated dynamically when jobs are running. This should be tested for accessibility. Depending on the importance of updates, they should be read out to the user straight away, as soon as possible, when the user is idle, or not at all.
- [ ] Check that the tab order makes sense.

**Comment by mccallum-sgd** [comment]
> they should be read out to the user straight away

I'm not exactly sure how this can be implemented without just moving the focus to the appropriate node, and I think that would be quite annoying for the user. Any ideas?

**Comment by mccallum-sgd** [comment]
For the labels in GridPaneHelper, originally I had just set the [accessibleText](https://docs.oracle.com/javase/8/javafx/api/javafx/scene/Node.html#accessibleTextProperty) of all of the control nodes to their corresponding labels (which is in most cases the answer's name), and set the tooltip to the answer's description (which also sets [accessibleHelp](https://docs.oracle.com/javase/8/javafx/api/javafx/scene/Node.html#accessibleHelpProperty) automatically). This seemed to work well for short descriptions, but for long ones it became a bit much. 

So I then tried just making the labels and help text focusTraversable, which also works fine and I believe is more conventional. Which should I do? (maybe both? it couldn't hurt i suppose)

**Comment by bertfrees** [comment]
I'm not sure about making the option names and descriptions focusTraversable. Are you sure it's more conventional? I think labels and tooltips are the way to go. Having labels and tooltips and making the names and descriptions traversable at the same time just creates a lot of redundancy. Making everything traversable also forces the user to press TAB a lot more.

> This seemed to work well for short descriptions, but for long ones it became a bit much.

What do you mean by that exactly? And how does making the descriptions traversable solve this?

> Also it would be nice if the script's description was read out as the user navigates scriptsCombo.

That would indeed be nice. I've tried it with VoiceOver. But it does not really work. This is what VoiceOver says:

> DAISY 2.02 to EPUB 3 Choose a script: pop up button

...and after a few seconds:
> You are currently on a pop up button, inside of a scroll area. To display a list of options, press Control-Option-Space. The help tag is: Transforms a DAISY 2.02 publications into an EPUB 3 publication.

After pressing down arrow: the selection changes but nothing is read out.

After pressing Control-Option-Space:
> press DAISY 2.02 to EPUB 3 Choose a script: pop up button. Interact with list row 1 of 19 selected

...and after a few seconds:
> You are currently on a text element

After pressing down arrow:
> DAISY 2.02 Validator

...and after a few seconds:
> You are currently on a text element

**Comment by bertfrees** [comment]
I've tried adding this inside `newScriptSelected` but that didn't help.
```java
scriptsCombo.notifyAccessibleAttributeChanged(AccessibleAttribute.TEXT);
scriptsCombo.notifyAccessibleAttributeChanged(AccessibleAttribute.HELP);
```

**Comment by bertfrees** [comment]
Adding this to `ListCell`'s `updateItem` also didn't help:
```java
setAccessibleText(t.getName());
setAccessibleHelp(t.getDescription());
```

**Comment by mccallum-sgd** [comment]
> Are you sure it's more conventional?

I originally thought that making the text and textflows focusTraversable would allow them to be "explorable" by the screen reader (i.e. read first sentence, read next word, read next character etc.), but after actually testing it I realize this is only the case with editable text. So yes I think there's really no benefit to this and just using the accessibleText and accessibleHelp of the controls is the better solution.

#### ScriptsCombo behavior w/ VoiceOver:
That's pretty much what I experienced. I really don't know why my current solution doesn't work or either of yours. I remember I got it working a while ago, but it wasn't quite working with NVDA specifically so I tried other things. I'll have to see if I can recreate what I did because unfortunately I can't remember, but I believe I was only using setLabelFor at that point.

**Comment by bertfrees** [comment]
Prashant has tested the latest version and gives some suggestions:

1. The screen reader reads only the "name"  in the jobs list. It does not read the Status and ID. This field is announced by Jaws as a Table instead of a list. The data here is arranged like a table but the screen reader table commands cannot be supported. Therefore, it should be labelled as a list.

2. On my PC the text in the edit fields such as "Time tolerance", NCC path etc.   edit box was not being read. It may be further verified.

3. It will be good if a keyboard shortcut such as F6 is supported to move between the Jobs list, messages and  the main jobs details window.

**Comment by bertfrees** [comment]
Avneesh reported another issue:

> When I invoked new script by usin ctrl+N, it crashed.

... maybe not related to accessibility, but similar to something he mentioned in his original accessibility report:

> When I press browse button for output directory, the select directory dialog opens, and while tabbing in this select directory dialog something makes the GUI crash on my machine.

**Comment by bertfrees** [comment]
> The screen reader reads only the "name"  in the jobs list. It does not read the Status and ID. This field is announced by Jaws as a Table instead of a list. The data here is arranged like a table but the screen reader table commands cannot be supported. Therefore, it should be labelled as a list.

I fixed this in 4cdd2a818109afacb2daeb2c1501e82de47e565c

> On my PC the text in the edit fields such as "Time tolerance", NCC path etc.   edit box was not being read. It may be further verified.

I fixed this in 632d4831fd5fced2ece8a48e815309d3b8b0e534

> It will be good if a keyboard shortcut such as F6 is supported to move between the Jobs list, messages and  the main jobs details window. 

I implemented this in 46a1f73faaf4917ee0949f885fd4002f0c8dcf83.

**Comment by bertfrees** [comment]
Fixed in https://github.com/daisy/pipeline-gui/commit/5d8290273083f5d94900fc728e4358db513cdec3 (further accessibility improvements will be tracked in a new issue).
